src/System.Net.Http/src/System/Net/Http/HPack/Huffman.cs

```diff
@@ -0,0 +1,426 @@
+        {
+            return _encodingTable[data];
+        }
+
+        /// <summary>
+        /// Decodes a Huffman encoded string from a byte array.
+        /// </summary>
+        /// <param name="src">The source byte array containing the encoded data.</param>
+        /// <param name="offset">The offset in the byte array where the coded data starts.</param>
+        /// <param name="count">The number of bytes to decode.</param>
+        /// <param name="dst">The destination byte array to store the decoded data.</param>
+        /// <returns>The number of decoded symbols.</returns>
+        public static int Decode(byte[] src, int offset, int count, byte[] dst)
+        {
+            var i = offset;
+            var j = 0;
+            var lastDecodedBits = 0;
+            while (i < count)
+            {
+                var next = (uint)(src[i] << 24 + lastDecodedBits);
+                next |= (i + 1 < src.Length ? (uint)(src[i + 1] << 16 + lastDecodedBits) : 0);
+                next |= (i + 2 < src.Length ? (uint)(src[i + 2] << 8 + lastDecodedBits) : 0);
+                next |= (i + 3 < src.Length ? (uint)(src[i + 3] << lastDecodedBits) : 0);
+
+                var ones = (uint)(int.MinValue >> (8 - lastDecodedBits - 1));
+                if (i == count - 1 && lastDecodedBits > 0 && (next & ones) == ones)
+                {
+                    // The remaining 7 or less bits are all 1, which is padding.
+                    // We specifically check that lastDecodedBits > 0 because padding
+                    // longer than 7 bits should be treated as a decoding error.
+                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                    break;
+                }
+
+                // The longest possible symbol size is 30 bits. If we're at the last 4 bytes
+                // of the input, we need to make sure we pass the correct number of valid bits
+                // left, otherwise the trailing 0s in next may form a valid symbol.
+                var validBits = Math.Min(30, (8 - lastDecodedBits) + (count - i - 1) * 8);
+                var ch = Decode(next, validBits, out var decodedBits);
+
+                if (ch == -1)
+                {
+                    // No valid symbol could be decoded with the bits in next
+                    throw new HuffmanDecodingException("CoreStrings.HPackHuffmanErrorIncomplete");
+                }
+                else if (ch == 256)
+                {
+                    // A Huffman-encoded string literal containing the EOS symbol MUST be treated as a decoding error.
+                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                    throw new HuffmanDecodingException("CoreStrings.HPackHuffmanErrorEOS");
+                }
+
+                if (j == dst.Length)
+                {
+                    throw new HuffmanDecodingException("CoreStrings.HPackHuffmanErrorDestinationTooSmall");
```

Comment:
Going back to the same error reporting story, we probably don't need or want a separate exception type for this... it could just be an error code in whatever top-level exception we throw.